pisa/stages/xsec/correct_charm_y.py

```diff
@@ -0,0 +1,163 @@
+
+from pisa.core.stage import Stage
+from pisa.utils.resources import open_resource
+from pisa.utils import vectorizer
+from pisa.utils.fileio import from_file
+from pisa.utils.log import logging
+from pisa import FTYPE, TARGET
+
+class correct_charm_y(Stage):
+    """
+    This stage reweights inelasticity distribution
+    to remove the most obvious impact from charm
+    production bug in GENIE 2.12.8. The bug effectively 
+    kills most charm events.
+
+    Parameters:
+    -----------
+    correct_charm : bool
+      If true, weights will be updated to include reweighting
+      factor correcting inelasticity distribution to follow one
+      we would have if we had charm in our MC.
+
+
+    Notes:
+    ------
+    THIS STAGE IS FOR MC TESTING ONLY - do *NOT* use in real analysis!!!
+
+
+    References
+    ----------
+    Slides summarizing the bug (osc. call, 28 Nov 2022): 
+    https://drive.google.com/file/d/16tF_ofjI-YTxnxSsuZD4W2X52xND1CAM/view?usp=drive_link
+
+    Slides showing use for check with oscNext VS (osc. call, 19 Dec 2022): 
+    https://drive.google.com/file/d/14h0WYWPWn7yS7wLQpyMAPTenfKfFAq3k/view?usp=drive_link
+
+    """
+    def __init__(
+            self,
+            correct_charm=False,
+            **std_kwargs,
+    ):
+
+        self.correct_charm = correct_charm
+        expected_params = ()
+
+        expected_container_keys = (
+            'true_energy',
+            'true_coszen',
+            'bjorken_y',
+            'weights',
+        )
+
+        # init base class
+        super(correct_charm_y, self).__init__(
+            expected_params=expected_params,
+            expected_container_keys=expected_container_keys,
+            **std_kwargs,
+        )
+
+    def setup_function(self):
+
+        if self.correct_charm:
+            init_y_disrt_hists = from_file('cross_sections/charm_y_correction_2d_coszen_split.pckl')
+
+            h_nucc_upg = init_y_disrt_hists['nu_cc_upg']
+            h_nubarcc_upg = init_y_disrt_hists['nubar_cc_upg']
+            h_nucc_oth = init_y_disrt_hists['nu_cc_oth']
+            h_nubarcc_oth = init_y_disrt_hists['nubar_cc_oth']
+            xed = init_y_disrt_hists['bins_lgE']
+            yed = init_y_disrt_hists['bins_y']
+
+            def eval_hist(lgE, y, coszen, nubar=False):
+
+                ind_x = np.digitize(lgE, bins=xed)
+                ind_y = np.digitize(y, bins=yed)
+
+                # TODO: replace hardcoded number of bins?
+                ind_x[ind_x==31] = 30
+                ind_y[ind_y==31] = 30
+
+                res = np.zeros_like(lgE)
+                mask_upg = coszen<-0.9
+
+#                 if coszen<-0.9:
+#                     h_nucc = h_nucc_upg
+#                     h_nubarcc = h_nubarcc_upg
+#                 else:
+#                     h_nucc = h_nucc_oth
+#                     h_nubarcc = h_nubarcc_oth
+
+                if not nubar:
+                    res[mask_upg] = h_nucc_upg[ind_x[mask_upg]-1, ind_y[mask_upg]-1]
```

Comment:
Why is a different correction applied to upgoing events?

Reply:
This was a later modification to the test, which was introduced to account for the fact that some events in the very upgoing region manage to survive physical momentum check in genie. Because of this inelasticity distribution is affected differently in very upgoing events and a different correction ratio was produced.